text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+            │                                          │
+            ↓                                          ↓
+┌───────────────────────┐             ┌────────────────────────────────────┐
+│ const ValueSized      │ ──────────→ │ ValueSized                         │
+│ {type, target, value} │   implies   │ {type, target, runtime env, value} │
+└───────────────────────┘             └────────────────────────────────────┘
+                                                       │
+                                                    implies
+                                                       │
+                               ┌───────────────────────┘
+                               ↓
+                     ┌──────────────────┐
+                     │ Pointee          │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+                               │
+                            implies
+                               │
+                               ↓
+                     ┌──────────────────┐
+                     │ Value            │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+```
+
+`Pointee` is still defined as being implemented for any type that can be used
+behind a pointer and may not be sized at all, this would be implemented for
+effectively every type except wasm's `externref` (or similar opaque types from
+some GPU targets). `Value` is defined as being implemented for any type that can
+be used as a value, which is all types, and also may not be sized at all.
+
+Earlier in this RFC, `extern type`s have previously been described as not being
+able to be used as a value, but it could instead be permitted to write functions
+which use extern types as values (e.g. such as taking an extern type as an argument),
+and instead rely on it being impossible to get a extern type that is not behind a
+pointer or a reference. This also implies that `ValueSized` types can be used as values,
+which would remain prohibited behind the `unsized_locals` and `unsized_fn_params`
+features until these are stabilised.
+
+With these changes to the RFC, it would be possible to support wasm's `externref` and
+opaque types from some GPU targets.
+
+## Alignment
+[alignment]: #alignment
+
+There has been community interest in an [`Aligned` trait][rfc_aligned] and there
+are examples of `Aligned` traits being added in the ecosystem:
+
+- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
+- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
+  whether a type has an alignment or not.
+
+Furthermore, the existing incorrect behaviour of `align_of_val` returning one is
+used in `rustc` with the [`OpaqueListContents` type][rustc_opaquelistcontents] to
+implement a custom DST. This use case would break with this RFC as written, as
+the alignment of an extern type would correctly be undefined.
```

Comment:
This is not correct, since align_of_val does not return 1 any more for extern types.

Reply:
Fixed!